src/precompiles/modexp.rs

```diff
@@ -3,107 +3,41 @@ use crate::precompiles::{
 };
 use crate::prelude::{PhantomData, Vec, U256};
 use evm::{Context, ExitError};
-use num::BigUint;
+use num::{BigUint, Integer};
 
 pub(super) const ADDRESS: [u8; 20] = super::make_address(0, 5);
 
 pub(super) struct ModExp<HF: HardFork>(PhantomData<HF>);
 
-impl ModExp<Byzantium> {
-    fn adj_exp_len(exp_len: U256, base_len: U256, bytes: &[u8]) -> U256 {
-        let mut exp32_bytes = Vec::with_capacity(32);
-        for i in 0..32 {
-            if U256::from(96) + base_len + U256::from(1) >= U256::from(bytes.len()) {
-                exp32_bytes.push(0u8);
-            } else {
-                let base_len_i = base_len.as_usize();
-                let bytes_i = 96 + base_len_i + i;
-                if let Some(byte) = bytes.get(bytes_i) {
-                    exp32_bytes.push(*byte);
-                } else {
-                    // Pad out the data if the byte is empty.
-                    exp32_bytes.push(0u8);
-                }
-            }
-        }
-        let exp32 = U256::from(exp32_bytes.as_slice());
+impl<HF: HardFork> ModExp<HF> {
+    fn calc_iter_count(exp_len: u64, base_len: u64, bytes: &[u8]) -> u64 {
+        let start = base_len as usize + 96;
+        let end = exp_len as usize + start;
```

Comment:
```suggestion
        let end = cmp::min(32, exp_len as usize) + start;
```

We only want the first 32 bytes of the exponent for this function.